### PR TITLE
Add Dialog::run_future()

### LINKED ISSUE #144: Keep gtk::DialogExt::run() as async variant
When looking at my usage of `dialog.run()` I decided to just implement an `async run()`.

The only solution I see is to connect to the "response" event and send the result through a channel. But I can also check how gtk3 implemented this.

Any thoughts?

### PR DESCRIPTION
**Pull request by sophie-h**
Closes #144

### CODE REVIEW

File: examples/src/bin/dialog.rs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+async fn dialog<W: IsA<gtk::Window>>(window: Rc<W>) {
+    let dialog = gtk::MessageDialog::new(
+        Some(&*window),
+        gtk::DialogFlags::empty(),
+        gtk::MessageType::Question,
+        gtk::ButtonsType::OkCancel,
+        "Hello",
+    );
+    println!("{:#?}", dialog.run_future().await);
```
Comment (sophie-h) [inline examples/src/bin/dialog.rs:22]: Replace with reaction in interface

File: examples/Cargo.toml [hunk 2]
```diff
@@ -3,47 +3,10 @@ name = "gtk4-rs-examples"
...
 
 [dependencies]
 chrono = "0.4.19"
 
 [dependencies.gtk]
 path = "../gtk4"
 package = "gtk4"
-
-[[bin]]
-name = "basic"
```
Comment (sdroege) [inline examples/Cargo.toml:16]: Can you put this into a separate commit?

File: gtk4/Cargo.toml [hunk 3]
```diff
@@ -37,9 +37,11 @@ optional = true
 git = "https://github.com/gtk-rs/lgpl-docs"
 
 [dependencies]
+async-trait = "0.1"
```
Comment (sdroege) [inline gtk4/Cargo.toml:40]: Let's do this manually for now instead of adding another dependency. It's easy enough

File: gtk4/src/dialog.rs [hunk 4]
```diff
@@ -30,15 +33,19 @@ impl Dialog {
...
 
+#[async_trait(?Send)]
 pub trait DialogExtManual: 'static {
     #[doc(alias = "gtk_dialog_add_buttons")]
     fn add_buttons(&self, buttons: &[(&str, ResponseType)]);
 
     #[doc(alias = "gtk_dialog_get_response_for_widget")]
     fn get_response_for_widget<P: IsA<Widget>>(&self, widget: &P) -> ResponseType;
+
+    async fn run_future(&self) -> ResponseType;
```
Comment (sdroege) [inline gtk4/src/dialog.rs:44]:
```suggestion
    fn run_future(&self) -> impl Future<Output = ResponseType>;
```

File: gtk4/src/dialog.rs [hunk 5]
```diff
@@ -49,9 +56,31 @@ impl<O: IsA<Dialog>> DialogExtManual for O {
...
+        let sender = std::cell::Cell::new(Some(sender));
+
+        let response_handler = self.connect_response(move |_, response_type| {
+            if let Some(m) = sender.replace(None) {
+                let _result = m.send(response_type);
+            }
+        });
+
+        let modal = self.get_modal();
+        self.set_modal(true);
```
Comment (sdroege) [inline gtk4/src/dialog.rs:77]: Why set it to modal?

File: gtk4/src/dialog.rs [hunk 6]
```diff
@@ -49,9 +56,31 @@ impl<O: IsA<Dialog>> DialogExtManual for O {
...
         unsafe {
             from_glib(ffi::gtk_dialog_get_response_for_widget(
-                self.as_ref().to_glib_none().0,
+                AsRef::<Dialog>::as_ref(self).to_glib_none().0,
                 widget.as_ref().to_glib_none().0,
             ))
         }
     }
+
+    async fn run_future(&self) -> ResponseType {
```
Comment (sdroege) [inline gtk4/src/dialog.rs:65]:
```suggestion
    fn run_future(&self) -> Pin<Box<Future<Output = ResponseType>>> {
```

Comment (sdroege) [reply gtk4/src/dialog.rs:65]: And the body would basically be

```
async {
    // What you have now
}
```

File: gtk4/Cargo.toml (hunk 3 above)
Comment (sophie-h) [reply gtk4/Cargo.toml]: Done

File: gtk4/src/dialog.rs [hunk 7]
```diff
@@ -49,9 +53,30 @@ impl<O: IsA<Dialog>> DialogExtManual for O {
...
+
+            let sender = std::cell::Cell::new(Some(sender));
+
+            let response_handler = self.connect_response(move |_, response_type| {
+                if let Some(m) = sender.replace(None) {
+                    let _result = m.send(response_type);
+                }
+            });
+
+            self.show();
```
Comment (sophie-h) [inline gtk4/src/dialog.rs:74]: Do we want to keep this? It doesn't seem too unreasonable.

Comment (sdroege) [reply gtk4/src/dialog.rs:74]: I'd make it behave like `gtk_dialog_run()` in GTK3. Least amount of surprises :) I don't know what exactly it did back then

File: gtk4/src/dialog.rs [hunk 8]
```diff
@@ -49,9 +53,30 @@ impl<O: IsA<Dialog>> DialogExtManual for O {
...
+
+            let result = receiver.await.unwrap_or(ResponseType::None);
+            self.disconnect(response_handler);
```
Comment (sdroege) [inline gtk4/src/dialog.rs:77]: If `receiver.await` returns an `Err`, then the signal handler was freed before the dialog got a response. In that case, `disconnect()` will probably give a critical warning so you probably only want to call it in the other case

**Review by sdroege — COMMENTED** [review]
Looks good to me otherwise, seems like a good usability improvement to me :)

**Comment by sdroege** [comment]
@bilelmoussaoui (doesn't show you in the reviewers list for some reason) Also seems like a good idea to you?

**Review by GuillaumeGomez — APPROVED** [review]

File: gtk4/src/dialog.rs (hunk 8 above)
Comment (sophie-h) [reply gtk4/src/dialog.rs]: Fixed.

**Comment by bilelmoussaoui** [comment]
> @bilelmoussaoui (doesn't show you in the reviewers list for some reason) Also seems like a good idea to you?

It feels a bit odd to me to start providing async variants on top of the basic bindings like this :)

**Comment by sdroege** [comment]
> It feels a bit odd to me to start providing async variants on top of the basic bindings like this :)

We also have those in glib/gio already though.

**Comment by bilelmoussaoui** [comment]
> > It feels a bit odd to me to start providing async variants on top of the basic bindings like this :)
> 
> We also have those in glib/gio already

**Comment by sophie-h** [comment]
> It feels a bit odd to me to start providing async variants on top of the basic bindings like this :)

Sounds like an argument that would also apply to all the *Builder traits that are also rusty variants on basic bindings. Also, I think that `run()` would still be part of GTK4 if it wasn't for the blocking behavior.

**Comment by bilelmoussaoui** [comment]
> > It feels a bit odd to me to start providing async variants on top of the basic bindings like this :)
> 
> Sounds like an argument that would also apply to all the *Builder traits that are also rusty variants on basic bindings. Also, I think that `run()` would still be part of GTK4 if it wasn't for the blocking behavior.

Not really, they are just a way of calling `glib::Object::new(&[])` =) I'm not really against such change, I was just wondering if/where we should draw a line between providing just bare-bone bindings or more than that.

**Comment by sdroege** [comment]
IMHO this is still fine, especially as it's not a lot of code and improves usability considerably. And it's not very different from the various futures API we provide in glib and gio already.

A function like this for gtk3 would also make sense.

@GuillaumeGomez What do you think?

**Comment by GuillaumeGomez** [comment]
My issue with this is a bit different: could we make this change automatic through gir? Could we detect the blocking `run` functions so we don't have generate them by hand?

**Comment by bilelmoussaoui** [comment]
> My issue with this is a bit different: could we make this change automatic through gir? Could we detect the blocking `run` functions so we don't have generate them by hand?

Gtk4 doesn't have any run functions anymore

**Comment by sdroege** [comment]
> Gtk4 doesn't have any run functions anymore

And for gtk3 we should really get rid of them. Re-entrant mainloops are a disaster :)

**Comment by GuillaumeGomez** [comment]
> And for gtk3 we should really get rid of them. Re-entrant mainloops are a disaster :)

Do we have an issue for this?

> Gtk4 doesn't have any run functions anymore

Then how can we add a `run_future`? I'm lost here...

**Comment by sdroege** [comment]
> Do we have an issue for this?

I don't think so but I changed all examples a long time ago.

> Then how can we add a `run_future`? I'm lost here...

By... implementing something that behaves like `run()` before :P Just look at the code

File: examples/src/bin/dialog.rs [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+    let info_dialog = gtk::MessageDialogBuilder::new()
+        .transient_for(&*window)
+        .modal(true)
+        .buttons(gtk::ButtonsType::Close)
+        .text("You answered")
+        .secondary_text(&format!("Your answer: {:?}", answer))
+        .build();
+
+    info_dialog.run_future().await;
```
Comment (sdroege) [inline examples/src/bin/dialog.rs:31]: Might make sense to do something with the return value here for example purposes

**Review by sdroege — APPROVED** [review]

**Comment by sophie-h** [comment]
> And for gtk3 we should really get rid of them. Re-entrant mainloops are a disaster :)

We could deprecate `run()` in gtk3-rs and add the future version (only requires extra handling of the delete event.)

File: examples/src/bin/dialog.rs (hunk 9 above)
Comment (sophie-h) [reply examples/src/bin/dialog.rs:31]: The answer from the previous dialog is shown. I think this is a good example why you might wan't to use async/await even if it's only for closing the dialog afterwards.

Comment (sdroege) [reply examples/src/bin/dialog.rs:31]: ok :)

**Comment by sdroege** [comment]
Now just need to fix the CI. Something needs to be done to allow `await` inside documentation. @GuillaumeGomez ?

**Comment by GuillaumeGomez** [comment]
After thinking about this for a while, I think it'd be nice to add it. **However**: please add the non-future equivalent (with a callback).

**Comment by GuillaumeGomez** [comment]
> Something needs to be done to allow `await` inside documentation. @GuillaumeGomez ?

Just wrap it inside an async function?

**Comment by sdroege** [comment]
> After thinking about this for a while, I think it'd be nice to add it. **However**: please add the non-future equivalent (with a callback).

Why? That already exists: connect to the `response` signal.

**Comment by GuillaumeGomez** [comment]
Then I can answer the same for `run_future`. :)

**Comment by sdroege** [comment]
> Then I can answer the same for `run_future`. :)

Signals can't generally be used as futures, that's why such a helper API here is useful for this specific signal.

**Comment by GuillaumeGomez** [comment]
Maybe, but it can be done as proved by this PR. So the same can be done for `run`. It wouldn't make sense to only have future version of `run`.

**Comment by sdroege** [comment]
> Maybe, but it can be done as proved by this PR. So the same can be done for `run`. It wouldn't make sense to only have future version of `run`.

Can you implement `run()` with a closure then? I don't know how you imagine that to look like

**Comment by GuillaumeGomez** [comment]
Sure.

File: gtk4/src/dialog.rs [hunk 10]
```diff
@@ -36,9 +38,24 @@ pub trait DialogExtManual: 'static {
 
     #[doc(alias = "gtk_dialog_get_response_for_widget")]
     fn get_response_for_widget<P: IsA<Widget>>(&self, widget: &P) -> ResponseType;
+
+    /// Shows the dialog and returns a `Future` that resolves to the
+    /// `ResponseType` on response.
+    ///
+    /// ```
```
Comment (GuillaumeGomez) [inline gtk4/src/dialog.rs:45]: Please add `no_run`.

File: gtk4/src/dialog.rs [hunk 11]
```diff
@@ -36,9 +38,24 @@ pub trait DialogExtManual: 'static {
 
     #[doc(alias = "gtk_dialog_get_response_for_widget")]
     fn get_response_for_widget<P: IsA<Widget>>(&self, widget: &P) -> ResponseType;
+
+    /// Shows the dialog and returns a `Future` that resolves to the
```
Comment (GuillaumeGomez) [inline gtk4/src/dialog.rs:42]: Please add before this doc comment:

```rust
// rustdoc-stripper-ignore-next
```

File: gtk4/src/dialog.rs (hunk 10 above)
Comment (sophie-h) [reply gtk4/src/dialog.rs]: Thanks! Fixed.

File: gtk4/src/dialog.rs (hunk 11 above)
Comment (sophie-h) [reply gtk4/src/dialog.rs:44]: Thanks! Fixed.

**Comment by sophie-h** [comment]
All green
